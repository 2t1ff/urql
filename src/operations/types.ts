import { DocumentNode, OperationDefinitionNode } from 'graphql';
import { FragmentSelectionSets, VarsMap } from '../ast';
import { Store } from '../store';
import { Scalar, SystemFields } from '../types';

export interface DataFields {
  [fieldName: string]: Data | Data[] | Scalar;
}

export type Data = DataFields & SystemFields;

export interface Request {
  query: DocumentNode;
  variables?: null | object;
}

export interface Result {
  data?: null | Data;
  isComplete: boolean;
  dependencies: string[];
}

export interface Context {
  /** List of heuristic keys indicating what paths have been walked */
  dependencies: string[];
  /** When reading from cache it indicates whether the result is likely complete */
  isComplete: boolean;
  operation: OperationDefinitionNode;
  store: Store;
  fragments: FragmentSelectionSets;
  vars: VarsMap;
}
