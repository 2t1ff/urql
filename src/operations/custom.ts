import { FieldResolver } from '../types';

export const makeCustomResolver = (
  innerResolver: FieldResolver
): FieldResolver => {
  return (fieldName, rootValue, args, context, info) => {
    const { __typename: typeName } = rootValue;
    const fieldResolver = context.store.getResolver(typeName, fieldName);
    if (fieldResolver !== undefined) {
      const result = fieldResolver(args, context, info);
      if (result !== undefined) {
        return result;
      }
    }

    return innerResolver(fieldName, rootValue, args, context, info);
  };
};
