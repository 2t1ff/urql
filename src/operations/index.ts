export { query } from './query';
export { write } from './write';
export { gc } from './gc';
