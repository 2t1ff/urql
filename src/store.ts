import { Entity, EntityMap, KeyExtractor, Link, LinkMap } from './types';
import {
  assignObjectToMap,
  isOperation,
  keyOfEntity,
  objectOfMap,
} from './utils';

export interface StoreData {
  records: {
    [key: string]: Entity;
  };
  links: {
    [key: string]: Link;
  };
}

export interface StoreOpts {
  initial?: StoreData;
  keyExtractor?: KeyExtractor;
}

class Store {
  private touched: string[];
  private records: EntityMap;
  private links: LinkMap;

  public keyOfEntity: (entity: Entity) => null | string;

  constructor({ initial, keyExtractor }: StoreOpts) {
    this.touched = [];
    this.records = new Map();
    this.links = new Map();

    if (keyExtractor !== undefined) {
      this.keyOfEntity = (entity: Entity) => {
        const key = keyExtractor(entity);
        return key !== undefined ? key : keyOfEntity(entity);
      };
    } else {
      this.keyOfEntity = keyOfEntity;
    }

    if (initial !== undefined) {
      assignObjectToMap(this.records, initial.records);
      assignObjectToMap(this.links, initial.links);
    }
  }

  getEntity(key: string): null | Entity {
    if (!isOperation(key)) {
      this.touched.push(key);
    }

    const entity = this.records.get(key);
    return entity !== undefined ? entity : null;
  }

  getOrCreateEntity(key: string): Entity {
    const prev = this.records.get(key);
    if (prev !== undefined && prev !== null) {
      return prev;
    }

    const entity = Object.create(null);
    this.records.set(key, entity);
    return entity;
  }

  writeEntityValue(key: string, prop: string, val: any) {
    if (!isOperation(key)) {
      this.touched.push(key);
    }

    const entity = this.getOrCreateEntity(key);
    if (val === undefined) {
      delete entity[prop];
    } else {
      entity[prop] = val;
    }
  }

  getLink(key: string): Link {
    this.touched.push(key);
    const link = this.links.get(key);
    return link !== undefined ? link : null;
  }

  writeLink(key: string, link: Link) {
    this.touched.push(key);

    if (link === null) {
      this.links.delete(key);
    } else {
      this.links.set(key, link);
    }
  }

  toJSON(): StoreData {
    return {
      records: objectOfMap(this.records),
      links: objectOfMap(this.links),
    };
  }

  flushTouched(): string[] {
    const touched = this.touched.filter((key, i, arr) => {
      return arr.indexOf(key) === i;
    });

    this.touched = [];
    return touched;
  }
}

export default Store;
