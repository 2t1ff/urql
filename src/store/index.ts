import { Entity, Link, Links, Records } from '../types';
import { assignObjectToMap, objectOfMap } from './utils';

export interface Store {
  records: Records;
  links: Links;
}

export interface SerializedStore {
  records: { [link: string]: Entity };
  links: { [link: string]: Link };
}

/** Creates a new Store with an optional initial, serialized state */
export const create = (initial?: SerializedStore): Store => {
  const records: Records = new Map();
  const links: Links = new Map();

  if (initial !== undefined) {
    assignObjectToMap(records, initial.records);
    assignObjectToMap(links, initial.links);
  }

  return { records, links };
};

/** Serializes a given Store to a plain JSON structure */
export const serialize = (store: Store): SerializedStore => {
  const records = objectOfMap(store.records);
  const links = objectOfMap(store.links);
  return { records, links };
};

export const find = (store: Store, key: string): Entity | null => {
  const entity = store.records.get(key);
  return entity !== undefined ? entity : null;
};

export const findOrCreate = (store: Store, key: string): Entity => {
  const entity = find(store, key);
  if (entity !== null) {
    return entity;
  }

  const record: Entity = Object.create(null);
  store.records.set(key, record);
  return record;
};

export const setLink = (store: Store, key: string, link: Link): void => {
  store.links.set(key, link);
};

export const deleteLink = (store: Store, key: string): void => {
  store.links.delete(key);
};

export const readLink = (store: Store, key: string): void | Link =>
  store.links.get(key);
