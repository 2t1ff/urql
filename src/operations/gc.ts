import { joinKeys } from '../helpers';
import { find, readLink, remove, removeLink, Store } from '../store';
import { Link } from '../types';

interface Context {
  store: Store;
  visitedLinks: Set<string>;
  visitedRecords: Set<string>;
}

export const gc = (store: Store) => {
  const visitedLinks = new Set<string>();
  const visitedRecords = new Set<string>();
  const ctx = { store, visitedLinks, visitedRecords };

  walkEntity(ctx, 'query');

  store.records.forEach((_entity, key) => {
    if (!visitedRecords.has(key)) {
      remove(store, key);
    }
  });

  store.links.forEach((_link, key) => {
    if (!visitedLinks.has(key)) {
      removeLink(store, key);
    }
  });
};

const walkEntity = (ctx: Context, key: string) => {
  const { store, visitedRecords, visitedLinks } = ctx;
  const entity = find(store, key);

  if (entity !== null && !visitedRecords.has(key)) {
    visitedRecords.add(key);

    for (const fieldKey in entity) {
      const value = entity[key];
      if (value === null) {
        const childFieldKey = joinKeys(key, fieldKey);
        const link = readLink(store, childFieldKey);
        if (link !== undefined && !visitedLinks.has(childFieldKey)) {
          visitedLinks.add(childFieldKey);
          walkLink(ctx, link);
        }
      }
    }
  }
};

const walkLink = (ctx: Context, link: Link) => {
  if (Array.isArray(link)) {
    link.forEach(childLink => walkLink(ctx, childLink));
  } else if (link !== null) {
    walkEntity(ctx, link);
  }
};
