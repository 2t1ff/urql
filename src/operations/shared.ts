import { FieldNode } from 'graphql';

import {
  getFragments,
  getMainOperation,
  getName,
  getNormalizedVars,
  getSelectionSet,
  isFieldNode,
  isInlineFragment,
  SelectionSet,
  shouldInclude,
} from '../ast';

import { Store } from '../store';
import { Context, Request } from './types';

export const makeContext = (store: Store, request: Request): void | Context => {
  const { query, variables } = request;
  const operation = getMainOperation(query);
  if (operation === undefined) {
    return;
  }

  const dependencies = [];
  const fragments = getFragments(query);
  const vars = getNormalizedVars(operation, variables);
  const isComplete = true;

  return { dependencies, isComplete, operation, fragments, vars, store };
};

export const forEachFieldNode = (
  ctx: Context,
  select: SelectionSet,
  cb: (node: FieldNode) => void
) => {
  const { vars, fragments } = ctx;

  select.forEach(node => {
    if (!shouldInclude(node, vars)) {
      // Directives instruct this node to be skipped
      return;
    } else if (!isFieldNode(node)) {
      // A fragment is either referred to by FragmentSpread or inline
      const def = isInlineFragment(node) ? node : fragments[getName(node)];

      if (def !== undefined) {
        const fragmentSelect = getSelectionSet(def);
        // TODO: Check for getTypeCondition(def) to match
        // Recursively process the fragments' selection sets
        forEachFieldNode(ctx, fragmentSelect, cb);
      }
    } else {
      cb(node);
    }
  });
};
