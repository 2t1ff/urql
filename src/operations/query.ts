import {
  getFieldAlias,
  getFieldArguments,
  getName,
  getSelectionSet,
  SelectionSet,
} from '../ast';

import { joinKeys, keyOfField } from '../helpers';
import { find, readLink, Store } from '../store';
import { Entity, Link } from '../types';

import { forEachFieldNode, makeContext } from './shared';
import { Context, Data, Request, Result } from './types';

/** Reads a request entirely from the store */
export const query = (store: Store, request: Request): Result => {
  const ctx = makeContext(store, request);
  if (ctx === undefined) {
    return { isComplete: false, dependencies: [] };
  }

  const select = getSelectionSet(ctx.operation);
  const data = readEntity(ctx, 'query', select);

  return {
    data,
    isComplete: ctx.isComplete,
    dependencies: ctx.dependencies,
  };
};

const readEntity = (
  ctx: Context,
  key: string,
  select: SelectionSet
): Data | null => {
  const { store } = ctx;
  const entity = find(store, key);
  if (entity === null) {
    // Cache Incomplete: A missing entity for a key means it wasn't cached
    ctx.isComplete = false;
    return null;
  } else if (key !== 'query') {
    ctx.dependencies.push(key);
  }

  const data = Object.create(null);
  readSelection(ctx, entity, key, data, select);
  return data;
};

const readSelection = (
  ctx: Context,
  entity: Entity,
  key: string,
  data: Data,
  select: SelectionSet
): void => {
  const { store, vars } = ctx;

  forEachFieldNode(ctx, select, node => {
    const fieldName = getName(node);
    // The field's key can include arguments if it has any
    const fieldKey = keyOfField(fieldName, getFieldArguments(node, vars));
    const fieldValue = entity[fieldKey];
    const fieldAlias = getFieldAlias(node);
    const childFieldKey = joinKeys(key, fieldKey);
    if (key === 'query') {
      ctx.dependencies.push(childFieldKey);
    }

    if (node.selectionSet === undefined || fieldValue !== null) {
      // Cache Incomplete: An undefined field value means it wasn't cached
      ctx.isComplete = fieldValue !== undefined;
      data[fieldAlias] = fieldValue === undefined ? null : fieldValue;
    } else {
      // null values mean that a field might be linked to other entities
      const { selections: fieldSelect } = node.selectionSet;
      const link = readLink(store, childFieldKey);

      // Cache Incomplete: A missing link for a field means it's not cached
      if (link === undefined) {
        ctx.isComplete = false;
        data[fieldAlias] = null;
      } else {
        data[fieldAlias] = readField(ctx, link, fieldSelect);
      }
    }
  });
};

const readField = (
  ctx: Context,
  link: Link,
  select: SelectionSet
): null | Data | Data[] => {
  if (Array.isArray(link)) {
    // @ts-ignore: Link cannot be expressed as a recursive type
    return link.map(childLink => readField(ctx, childLink, select));
  } else if (link === null) {
    return null;
  }

  return readEntity(ctx, link, select);
};
