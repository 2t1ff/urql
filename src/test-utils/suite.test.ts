import { DocumentNode } from 'graphql';
import gql from 'graphql-tag';
import { query, write } from '../operations';
import { create, serialize } from '../store';

interface TestCase {
  query: DocumentNode;
  variables?: any;
  data: any;
}

const expectCacheIntegrity = (testcase: TestCase) => {
  const store = create();
  const request = { query: testcase.query, variables: testcase.variables };
  const writeRes = write(store, request, testcase.data);
  const queryRes = query(store, request);
  expect(queryRes.data).toEqual(testcase.data);
  expect(queryRes.isComplete).toBe(true);
  expect(queryRes.dependencies).toEqual(writeRes.dependencies);
  const json = serialize(store);
  expect(json).toMatchSnapshot();
  expect(queryRes.dependencies).toMatchSnapshot();
  return json;
};

it('int on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        int
      }
    `,
    data: { int: 42 },
  });
});

it('json on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        json
      }
    `,
    // The `__typename` field should not mislead the cache
    data: { json: { __typename: 'Misleading', test: true } },
  });
});

it('nullable field on query', () => {
  expectCacheIntegrity({
    query: gql`
      {
        missing
      }
    `,
    data: { missing: null },
  });
});

it('int field with arguments on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        int(test: true)
      }
    `,
    data: { int: 42 },
  });

  expect(store.records.query).toMatchObject({
    'int({"test":true})': 42,
  });
});

it('non-keyable entity on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        item {
          name
        }
      }
    `,
    // This entity has no `__typename` field
    data: { item: { name: 'Test' } },
  });

  expect(store.links['query.item']).toBe('query.item');
  expect(store.records['query.item']).toMatchObject({ name: 'Test' });
});

it('non-IDable entity on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        item {
          __typename
          name
        }
      }
    `,
    // This entity has a `__typename` but no ID fields
    data: { item: { __typename: 'Item', name: 'Test' } },
  });

  expect(store.links['query.item']).toBe('query.item');
  expect(store.records['query.item']).toMatchObject({
    __typename: 'Item',
    name: 'Test',
  });
});

it('entity on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        item {
          __typename
          id
          name
        }
      }
    `,
    data: { item: { __typename: 'Item', id: '1', name: 'Test' } },
  });

  expect(store.links['query.item']).toBe('Item:1');
  expect(store.records['Item:1']).toMatchObject({
    __typename: 'Item',
    id: '1',
    name: 'Test',
  });
});

it('entity with arguments on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        item(test: true) {
          __typename
          id
          name
        }
      }
    `,
    data: { item: { __typename: 'Item', id: '1', name: 'Test' } },
  });

  expect(store.links['query.item({"test":true})']).toBe('Item:1');
  expect(store.records.query).toMatchObject({ 'item({"test":true})': null });
});

it('entity with Int-like ID on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        item {
          __typename
          id
          name
        }
      }
    `,
    // This is the same as above, but with a number on `id`
    data: { item: { __typename: 'Item', id: 1, name: 'Test' } },
  });

  expect(store.links['query.item']).toBe('Item:1');
  expect(store.links['query.item']).toBe('Item:1');
  expect(store.records.query.item).toBe(null);

  expect(store.records['Item:1']).toMatchObject({
    __typename: 'Item',
    id: 1,
    name: 'Test',
  });
});

it('entity list on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        items {
          __typename
          id
        }
      }
    `,
    data: {
      items: [{ __typename: 'Item', id: 1 }, { __typename: 'Item', id: 2 }],
    },
  });

  expect(store.links['query.items']).toEqual(['Item:1', 'Item:2']);
  expect(store.records.query.items).toBe(null);

  expect(store.records['Item:1']).toMatchObject({
    __typename: 'Item',
    id: 1,
  });

  expect(store.records['Item:2']).toMatchObject({
    __typename: 'Item',
    id: 2,
  });
});

it('nested entity list on query', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        items {
          __typename
          id
        }
      }
    `,
    data: {
      // This is the same as above, but with a nested array and added null values
      items: [
        { __typename: 'Item', id: 1 },
        [{ __typename: 'Item', id: 2 }, null],
        null,
      ],
    },
  });

  expect(store.links['query.items']).toEqual([
    'Item:1',
    ['Item:2', null],
    null,
  ]);
  expect(store.records.query.items).toBe(null);

  expect(store.records['Item:1']).toMatchObject({
    __typename: 'Item',
    id: 1,
  });

  expect(store.records['Item:2']).toMatchObject({
    __typename: 'Item',
    id: 2,
  });
});

it('embedded object on entity', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        item {
          __typename
          id
          author {
            __typename
            name
          }
        }
      }
    `,
    data: {
      item: {
        __typename: 'Item',
        id: 1,
        author: {
          __typename: 'Author',
          name: 'Stanley',
        },
      },
    },
  });

  expect(store.links['query.item']).toBe('Item:1');
  expect(store.links['Item:1.author']).toBe('Item:1.author');

  expect(store.records.query.item).toBe(null);
  expect(store.records['Item:1']).toMatchObject({
    __typename: 'Item',
    id: 1,
    author: null,
  });

  expect(store.records['Item:1.author']).toMatchObject({
    __typename: 'Author',
    name: 'Stanley',
  });
});

it('embedded object on entity', () => {
  const store = expectCacheIntegrity({
    query: gql`
      {
        item {
          __typename
          id
          author {
            __typename
            id
            name
          }
        }
      }
    `,
    data: {
      item: {
        __typename: 'Item',
        id: 1,
        author: {
          __typename: 'Author',
          id: 1,
          name: 'Stanley',
        },
      },
    },
  });

  expect(store.links['query.item']).toBe('Item:1');
  expect(store.links['Item:1.author']).toBe('Author:1');

  expect(store.records.query.item).toBe(null);
  expect(store.records['Item:1']).toMatchObject({
    __typename: 'Item',
    id: 1,
    author: null,
  });

  expect(store.records['Author:1']).toMatchObject({
    __typename: 'Author',
    id: 1,
    name: 'Stanley',
  });
});
