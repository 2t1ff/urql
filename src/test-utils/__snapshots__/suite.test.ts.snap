// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`embedded object on entity 1`] = `
Object {
  "links": Object {
    "Item:1.author": "Item:1.author",
    "query.item": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "author": null,
      "id": 1,
    },
    "Item:1.author": Object {
      "__typename": "Author",
      "name": "Stanley",
    },
    "query": Object {
      "item": null,
    },
  },
}
`;

exports[`embedded object on entity 2`] = `
Array [
  "query.item",
  "Item:1",
  "Item:1.author",
]
`;

exports[`embedded object on entity 3`] = `
Object {
  "links": Object {
    "Item:1.author": "Author:1",
    "query.item": "Item:1",
  },
  "records": Object {
    "Author:1": Object {
      "__typename": "Author",
      "id": 1,
      "name": "Stanley",
    },
    "Item:1": Object {
      "__typename": "Item",
      "author": null,
      "id": 1,
    },
    "query": Object {
      "item": null,
    },
  },
}
`;

exports[`embedded object on entity 4`] = `
Array [
  "query.item",
  "Item:1",
  "Author:1",
]
`;

exports[`entity list on query 1`] = `
Object {
  "links": Object {
    "query.items": Array [
      "Item:1",
      "Item:2",
    ],
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
    },
    "Item:2": Object {
      "__typename": "Item",
      "id": 2,
    },
    "query": Object {
      "items": null,
    },
  },
}
`;

exports[`entity list on query 2`] = `
Array [
  "query.items",
  "Item:1",
  "Item:2",
]
`;

exports[`entity on query 1`] = `
Object {
  "links": Object {
    "query.item": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": "1",
      "name": "Test",
    },
    "query": Object {
      "item": null,
    },
  },
}
`;

exports[`entity on query 2`] = `
Array [
  "query.item",
  "Item:1",
]
`;

exports[`entity with Int-like ID on query 1`] = `
Object {
  "links": Object {
    "query.item": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
      "name": "Test",
    },
    "query": Object {
      "item": null,
    },
  },
}
`;

exports[`entity with Int-like ID on query 2`] = `
Array [
  "query.item",
  "Item:1",
]
`;

exports[`entity with arguments on query 1`] = `
Object {
  "links": Object {
    "query.item({\\"test\\":true})": "Item:1",
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": "1",
      "name": "Test",
    },
    "query": Object {
      "item({\\"test\\":true})": null,
    },
  },
}
`;

exports[`entity with arguments on query 2`] = `
Array [
  "query.item({\\"test\\":true})",
  "Item:1",
]
`;

exports[`int field with arguments on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "query": Object {
      "int({\\"test\\":true})": 42,
    },
  },
}
`;

exports[`int field with arguments on query 2`] = `
Array [
  "query.int({\\"test\\":true})",
]
`;

exports[`int on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "query": Object {
      "int": 42,
    },
  },
}
`;

exports[`int on query 2`] = `
Array [
  "query.int",
]
`;

exports[`json on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "query": Object {
      "json": Object {
        "__typename": "Misleading",
        "test": true,
      },
    },
  },
}
`;

exports[`json on query 2`] = `
Array [
  "query.json",
]
`;

exports[`nested entity list on query 1`] = `
Object {
  "links": Object {
    "query.items": Array [
      "Item:1",
      Array [
        "Item:2",
        null,
      ],
      null,
    ],
  },
  "records": Object {
    "Item:1": Object {
      "__typename": "Item",
      "id": 1,
    },
    "Item:2": Object {
      "__typename": "Item",
      "id": 2,
    },
    "query": Object {
      "items": null,
    },
  },
}
`;

exports[`nested entity list on query 2`] = `
Array [
  "query.items",
  "Item:1",
  "Item:2",
]
`;

exports[`non-IDable entity on query 1`] = `
Object {
  "links": Object {
    "query.item": "query.item",
  },
  "records": Object {
    "query": Object {
      "item": null,
    },
    "query.item": Object {
      "__typename": "Item",
      "name": "Test",
    },
  },
}
`;

exports[`non-IDable entity on query 2`] = `
Array [
  "query.item",
  "query.item",
]
`;

exports[`non-keyable entity on query 1`] = `
Object {
  "links": Object {
    "query.item": "query.item",
  },
  "records": Object {
    "query": Object {
      "item": null,
    },
    "query.item": Object {
      "name": "Test",
    },
  },
}
`;

exports[`non-keyable entity on query 2`] = `
Array [
  "query.item",
  "query.item",
]
`;

exports[`nullable field on query 1`] = `
Object {
  "links": Object {},
  "records": Object {
    "query": Object {
      "missing": null,
    },
  },
}
`;

exports[`nullable field on query 2`] = `
Array [
  "query.missing",
]
`;
